python/ray/_private/package.py

```diff
@@ -0,0 +1,160 @@
+    RAY_PKG_PREFIX = "_ray_pkg_"
+    hash_val = None
+    if working_dir:
+        assert isinstance(working_dir, str)
+        assert Path(working_dir).exists()
+        hash_val = _xor_bytes(hash_val, _hash_modules(Path(working_dir)))
+    for module in modules or []:
+        assert inspect.ismodule(module)
+        hash_val = _xor_bytes(hash_val, _hash_modules(Path(module.__file__).parent))
+    return RAY_PKG_PREFIX + hash_val.hex() + ".zip" if hash_val else None
+
+def create_project_package(pkg_file: Path, working_dir: str, modules):
+    with ZipFile(pkg_file, "w") as zip_handler:
+        if working_dir:
+            # put all files in /path/working_dir into zip
+            working_path = Path(working_dir)
+            _zip_module(working_path, working_path, zip_handler)
+        for module in modules or []:
+            logger.info(module.__file__)
+            # we only take care of modules with path like this for now:
+            #    /path/module_name/__init__.py
+            # module_path should be: /path/module_name
+            module_path = Path(module.__file__).parent
+            _zip_module(module_path, module_path.parent, zip_handler)
+
+def parse_uri(pkg_uri: str) -> (Protocol, str):
+    uri = urlparse(pkg_uri)
+    protocol = Protocol(uri.scheme)
+    return (protocol, uri.netloc)
+
+def fetch_package(pkg_uri: str, pkg_file: Path):
+    (protocol, pkg_name) = parse_uri(pkg_uri)
+    if protocol in (Protocol.GCS, Protocol.PIN_GCS):
+        code = internal_kv._internal_kv_get(pkg_uri)
+        code = code or b""
+        pkg_file.write_bytes(code)
+        return len(code)
+    else:
+        raise NotImplementedError(f"Protocol {protocol} is not supported")
+
+def push_package(pkg_uri: str, pkg_file: Path):
+    (protocol, pkg_name) = parse_uri(pkg_uri)
+    if protocol in (Protocol.GCS, Protocol.PIN_GCS):
+        data = pkg_file.read_bytes()
+        internal_kv._internal_kv_put(pkg_uri, data)
+    else:
+        raise NotImplementedError(f"Protocol {protocol} is not supported")
+    return len(data)
+
+def package_exists(pkg_uri: str):
+    (protocol, pkg_name) = parse_uri(pkg_uri)
+    if protocol in (Protocol.GCS, Protocol.PIN_GCS):
+        return internal_kv._internal_kv_exists(pkg_uri)
+    else:
+        raise NotImplementedError(f"Protocol {protocol} is not supported")
+
+
+def update_runtime_env(job_config: JobConfig):
+    # For now, we only support local directory and packages
+    working_dir = job_config.runtime_env.get("working_dir")
+    required_modules = job_config.runtime_env.get("local_modules")
+
+    if (not job_config.runtime_env.get('working_dir_uri')) and \
+      (working_dir or required_modules):
+        pkg_name = get_project_package_name(working_dir, required_modules)
+        job_config.runtime_env["working_dir_uri"] = \
+          Protocol.GCS.value + "://" + pkg_name
+
+def init_runtime_env(is_driver: bool, job_config: JobConfig, pkg_uri: str):
```

Comment:
Please split this into two functions, one called by the driver, and one called by the worker.